[Merged by Bors] - feat(to_additive): `(dont_translate := ...)` option

JovanGerb opened the pull request:
This PR adds the `(dont_translate := ...)` option to `to_additive`. This helps with additivizing delarations that have some multiplicative structure that can't be translated, such as a `GroupWithZero`, `Ring` or `Field`. Testing with this feature showed that `firstMultiplicativeArg` needed to be improved. That is also done by this PR (it is renamed to `findMultiplicativeArg`).

In a future PR some automation may be added to infer these `dont_translate` arguments automatically based on the type of the declaration. For example if the type includes `GroupWithZero G`, then `G` will be automatically not translated.

This PR now also uses this feature to use `to_additive` for as much as possible of `MonoidAlgebra`.

[#mathlib4 > to_additive on AddMonoidWithOne](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/to_additive.20on.20AddMonoidWithOne)

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

JovanGerb committed:
feat(to_additive): `(dont_translate := ...)` option

JovanGerb committed:
add doc-strings

JovanGerb committed:
Merge branch 'master' into Jovan-to_additive-dont_translate

JovanGerb committed:
Merge branch 'master' into Jovan-to_additive-dont_translate

JovanGerb committed:
.

JovanGerb committed:
remove long file option

JovanGerb committed:
fix firstMultiplicativeArg

JovanGerb committed:
Merge branch 'master' into Jovan-to_additive-dont_translate

JovanGerb committed:
fix `firstMultiplicativeArg`

JovanGerb committed:
use in `MonoidAlgebra`

JovanGerb committed:
fix

JovanGerb committed:
missing doc-strings

JovanGerb committed:
fix `findMultiplicativeArg`

JovanGerb committed:
fix test

JovanGerb committed:
finish using `to_additive`

JovanGerb committed:
clean up comments

fpvandoorn reviewed (commented):
Few comments, but otherwise the changes to `to_additive` LGTM. I didn't look at the library changes.

This raises a question I considered before: some declarations now have multiple additive versions. In some lemmas, you can freely and independently choose whether or not to translate `A` and `B`, given 4 possible declarations. At some point we might want to consider a way to give multiple additivized versions to a declaration. 

I raise this, because with this feature we could actually declare any of them. But we can definitely ignore this issue/possible feature for a while.

JovanGerb committed:
improve comments

JovanGerb committed:
add test with `.proof_1`

JovanGerb commented on Mathlib/Tactic/ToAdditive/Frontend.lean:
```diff
@@ -695,19 +747,35 @@ def declUnfoldAuxLemmas (decl : ConstantInfo) : MetaM ConstantInfo := do
+In the future we would like that the presence of `MonoidAlgebra k G` will automatically
+flag `k` as a type to not be translated.
+-/
+def getDontTranslates (given : List Ident) (type : Expr) : MetaM (List Nat) := do
```
We could, but that would duplicate the code and run it twice (recall that this function may be beefed up in a follow up PR). The other issue is that, in principle, the binder names could be different in the forall (the type) and the lambda (the value). But of course the `dont` list should be the same for `applyReplacementForall` and `applyReplacementLambda`.

fpvandoorn replied:
That's a good reason

fpvandoorn commented on Mathlib/Tactic/ToAdditive/Frontend.lean:
```diff
@@ -801,8 +869,11 @@ partial def transformDeclAux
+    if src == pre then
+      updateDecl tgt srcDecl cfg.reorder cfg.dontTranslate
+    else
+      updateDecl tgt srcDecl [] []
```
Ah yes, I forgot we unfold automatically generated decls in the proof now. Then this shouldn't be an issue, I think.

The `to_additive_relevant_arg` issue sounds like an oversight.

JovanGerb committed:
revert new uses of `to_additive`

JovanGerb commented:
Ok, I have addressed the above comments, and I've split out the changes to the library into the new PR #29036.

The issue of `to_additive_relevant_arg` (and `to_additive_reorder`) not getting transferred automatically for `simps` (and for `._eq_1` and for `inductive`/`structure` fields), I will leave for some future PR.

bryangingechen commented on Mathlib/Tactic/ToAdditive/Frontend.lean:
```diff
@@ -92,8 +92,18 @@ the fifth argument before the third argument. This is used in `to_dual` to swap
 -/
 syntax toAdditiveReorderOption := &"reorder" " := " (num+),+
+/--
+`dont_translate := ...` takes a list of type variables that should not be considered
+for translation. For example in
```
```suggestion
`dont_translate := ...` takes a list of type variables (separated by spaces) that should not be 
considered for translation. For example in
```

Is it worth adding a test / example where we apply `dont_translate` to multiple variables?

bryangingechen reviewed (commented):
Thanks! Unless you're planning to work on it soon, recording the details of the issue of `to_additive_relevant_arg` / `to_additive_reorder` not being transferred as a TODO or repo issue might not be a bad idea.

bors d+

JovanGerb committed:
add test with multiple `dont_translate` arguments

JovanGerb commented:
bors r+
